src/vm/gdbjit.cpp

```diff
@@ -2223,117 +2223,95 @@ bool NotifyGdb::BuildSymbolTableSection(MemBuf& buf, PCODE addr, TADDR codeSize)
+
+    // Approximate length of single section name.
+    // Used only to reduce memory reallocations.
+    static const int SECT_NAME_LENGTH = 11;
+
+    if (!strBuf.Resize(SECT_NAME_LENGTH * (SectionNamesCount + thunks_count)))
     {
-        char name[256];
-        sprintf(name, ".thunk_%i", i);
-        strcpy(bufPtr, name);
-        bufPtr += strlen(name) + 1;
+        return false;
     }
-    return true;
-}
 
-/* Build the ELF section headers table */
-bool NotifyGdb::BuildSectionTable(MemBuf& buf)
-{
-    int thunks_count = SymbolCount - 1 - method.GetCount();
     Elf_Shdr* sectionHeaders = new (nothrow) Elf_Shdr[SectionNamesCount + thunks_count];
-    Elf_Shdr* pSh = sectionHeaders;
-
     if (sectionHeaders == nullptr)
     {
         return false;
     }
-    
-    /* NULL entry */
-    pSh->sh_name = 0;
-    pSh->sh_type = SHT_NULL;
-    pSh->sh_flags = 0;
-    pSh->sh_addr = 0;
-    pSh->sh_offset = 0;
-    pSh->sh_size = 0;
-    pSh->sh_link = SHN_UNDEF;
-    pSh->sh_info = 0;
-    pSh->sh_addralign = 0;
-    pSh->sh_entsize = 0;
-    
-    ++pSh;
-    /* fill section header data */
-    uint32_t sectNameOffset = 1;
-    for (int i = 1; i < SectionNamesCount; ++i, ++pSh)
+
+    sectBuf.MemPtr = reinterpret_cast<char*>(sectionHeaders);
+    sectBuf.MemSize = sizeof(Elf_Shdr) * (SectionNamesCount + thunks_count);
+
+    Elf_Shdr* pSh = sectionHeaders;
+    uint32_t sectNameOffset = 0;
+
+    // Fill section headers and names
+    for (int i = 0; i < SectionNamesCount + thunks_count; ++i, ++pSh)
     {
-        pSh->sh_name = sectNameOffset;
-        sectNameOffset += strlen(SectionNames[i]) + 1;
-        pSh->sh_type = Sections[i].m_type;
-        pSh->sh_flags = Sections[i].m_flags;
-        pSh->sh_addr = 0;
-        pSh->sh_offset = 0;
-        pSh->sh_size = 0;
-        pSh->sh_link = SHN_UNDEF;
-        pSh->sh_info = 0;
-        pSh->sh_addralign = 1;
-        if (strcmp(SectionNames[i], ".symtab") == 0)
-            pSh->sh_entsize = sizeof(Elf_Sym);
+        char thunkSectNameBuf[256]; // temporary buffer for .thunk_# section name
+        const char *sectName;
+
+        bool isThunkSection = i >= SectionNamesCount;
+        if (isThunkSection)
+        {
+            sprintf(thunkSectNameBuf, ".thunk_%i", i);
+            sectName = thunkSectNameBuf;
+        }
         else
-            pSh->sh_entsize = 0;
-    }
+        {
+            sectName = SectionNames[i];
+        }
 
-    for (int i = SectionNamesCount; i < SectionNamesCount + thunks_count; ++i, ++pSh)
-    {
-        char name[256];
-        sprintf(name, ".thunk_%i", i);
+        // Ensure that there is enough memory for section name,
+        // reallocate if necessary.
         pSh->sh_name = sectNameOffset;
-        sectNameOffset += strlen(name) + 1;
+        sectNameOffset += strlen(sectName) + 1;
+        if (sectNameOffset > strBuf.MemSize)
+        {
+            if (!strBuf.Resize(sectNameOffset))
```

Comment:
It would be good to give it some additional space instead of just exactly what's needed, otherwise once you cross the initial buffer size, you would keep resizing on every following loop iteration. You can consider using either to add a fixed value to the Resize parameter or a value that grows on each call to Resize (starting with e.g. 32, on the next Resize using 64, then 128, ...)